### DISCUSSION TITLE
How to debug program when there's a TUI?

### DISCUSSION THREAD

**Discussion by Squirrelcoding** [gyscos/cursive#639] [Q&A] [answered]
I'm fairly new to systems programming and I was wondering how you debug when there's a TUI where errors normally go. Normally the error would be logged to the standard output but the TUI is in the way! I don't know how to display errors somewhere else such as a log file, so can anybody help me out here?

TL:DR; Since the TUI is in the way, how can I see errors?

**Comment by gyscos** [comment, marked as answer]
Hi!

To debug, you have a few main options:
* Debug from inside the application itself. This means print messages to a popup layer or something. You can also use [cursive's own basic debug console](https://github.com/gyscos/cursive/blob/main/cursive/examples/debug_console.rs), or [flexi logger](https://github.com/deinstapel/cursive-flexi-logger-view), a third-party library with much more features.
* You can debug the application in gdb or similar debuggers like you would a non-TUI application. You can then set breakpoints and debug the logic step by step, looking at the values or variables, ...
* You can redirect the standard error stream to a file, and check this file from a separate terminal. To do that, you would do something like `cargo run 2> error_file.txt`, and `tail -F error_file.txt` in another terminal.
